File: Java/Sorting/patiencesort.java
import java.util.*;
import java.util.stream.Collectors;

class Stack extends ArrayList<Integer> implements Comparable<Stack> {
    @Override
    public int compareTo(Stack other) {
        return this.get(this.size() - 1).compareTo(other.get(other.size() - 1));
    }
}

public class patiencesort {
    public static List<Integer> patienceSort(List<Integer> collection) {
        List<Stack> stacks = new ArrayList<>();

        for (Integer element : collection) {
            Stack newStack = new Stack();
            newStack.add(element);
            int i = Collections.binarySearch(stacks, newStack);
            if (i >= 0) {
                stacks.get(i).add(element);
            } else {
                stacks.add(-(i + 1), newStack);
            }
        }

        List<Integer> sorted = stacks.stream()
                .flatMap(Collection::stream)
                .collect(Collectors.toList());

        return sorted;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.println("Enter numbers separated by a comma:");
        String userInput = scanner.nextLine().trim();
        List<Integer> unsorted = Arrays.stream(userInput.split(","))
                .map(Integer::parseInt)
                .collect(Collectors.toList());
        List<Integer> sorted = patienceSort(unsorted);
        System.out.println("Sorted Array:");
        System.out.println(sorted);
    }
}

File: Python/Sorting/Patience_sort.py
from __future__ import annotations

from bisect import bisect_left
from functools import total_ordering
from heapq import merge

@total_ordering
class Stack(list):
    def __lt__(self, other):
        return self[-1] < other[-1]

    def __eq__(self, other):
        return self[-1] == other[-1]


def patience_sort(collection: list) -> list:
   
    stacks: list[Stack] = []
    
    for element in collection:
        new_stacks = Stack([element])
        i = bisect_left(stacks, new_stacks)
        if i != len(stacks):
            stacks[i].append(element)
        else:
            stacks.append(new_stacks)

    collection[:] = merge(*(reversed(stack) for stack in stacks))
    return collection


if __name__ == "__main__":
    user_input = input("Enter numbers separated by a comma:\n").strip()
    unsorted = [int(item) for item in user_input.split(",")]
    print("sorted array:")
    print(patience_sort(unsorted))